### PR TITLE
many: introduce and use snap.SelfContainedSetPrereqTracker

### PR DESCRIPTION
**Pull request by pedronis**
also in snap.ValidateBasesAndProviders

these now can produce warnings, OTOH the relaxed checks allow to build/seed an
image even if a content requirement is fulfilled by an alternative provider

notice that with the relaxed checks seeding might fail or the system not work
if the right auto-connections or connections are not in-place

snaps used as default-providers usually have been taken care of already but it
might be up to the user to ask to set that up for alternative ones

### CODE REVIEW

File: snap/validate.go [hunk 1]
```diff
@@ -1188,15 +1188,114 @@ func NeededDefaultProviders(info *Info) (providerSnapsToContentTag map[string][]
...
+		prqt.all.Add(info)
+		prqt.snaps = append(prqt.snaps, info)
+	}
+}
+
+// DefaultProviderContentAttrs implements snapstate.PrereqTracker.
+// Given how snapstate uses this and as SelfContainedSetPrereqTracker is for
+// when no automatic fetching of prerequisites is desired, this always returns
+// nil.
+func (prqt *SelfContainedSetPrereqTracker) DefaultProviderContentAttrs(info *Info, repo InterfaceRepo) map[string][]string {
```
Comment (andrewphelpsj) [inline snap/validate.go:1222]: Could you explain the reasoning for returning `nil` here more? From looking at the doc comment on `DefaultProviderContentAttrs` in the `PrereqTracker` interface, I don't understand how other implementations besides the one in `SimplePrereqTracker` make sense.

**Review by andrewphelpsj — COMMENTED** [review]
Thank you! Just one question that I'm a little confused on.

File: snap/validate.go (hunk 1 above)
Comment (pedronis) [reply snap/validate.go:1222]: see the detail I added to https://github.com/snapcore/snapd/pull/13340/files#diff-9e832b3649d23b262a0cc4837c71c91a0e5d72ed1f2977cd7151b5a8e27da360R1097 

But I agree that maybe that comment needs to be changed further and the method needs a different name (the latter is a bit disruptive though at this point)

Comment (andrewphelpsj) [reply snap/validate.go:1222]: I see, that makes sense. Thanks!

Comment (pedronis) [reply snap/validate.go:1222]: ah, it is used indirectly in most of snapstate. So maybe renaming it to MissingProviderContentTags ? and changing a bit the doc comment in the interface?

Comment (andrewphelpsj) [reply snap/validate.go:1222]: I think that would help, yes.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Some minor comments around

File: seed/seedwriter/writer.go [hunk 2]
```diff
@@ -1287,6 +1288,13 @@ func (w *Writer) checkPrereqsInMode(mode string) error {
 		}
 		return fmt.Errorf("%s: %v", errPrefix, errs[0])
```
[the remark this answers is not in the document]
Comment (pedronis) [reply seed/seedwriter/writer.go:1289]: here I agree that we could do better in a follow up, I'll add a TODO

**Comment by pedronis** [comment]
@alfonsosanchezbeato @andrewphelpsj I applied some of your feedback, some will need to be addressed in a follow-up for 2.62

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for changes&responses

**Review by andrewphelpsj — APPROVED** [review]
Thanks for the rename and the expansion of the comments, that makes things more clear.

**Comment by ernestl** [comment]
I fixed a merge conflict caused by separate functions added in master and in this PR, by simply adding all.
https://github.com/snapcore/snapd/pull/13340/commits/c00972ae9a5f84452e60b85c65a5ba73061699e7

**Pull request merged**